### DISCUSSION TITLE
Using Windows (Hello) fingerprint with Webauthn

### DISCUSSION THREAD

**Discussion by prae82** [authelia/authelia#7522] [Q&A] [answered]
Hi all,

I have authelia running with Traefik, it works perfectly. I managed to enable TOTP and Webauthn too.
But for Webauthn I was only able to use my phones fingerprint/pin/etc. Not the ones which are available in Windows 11.

I know it is possible since I was using Webauthn in Proxmox before which did let me use the Windows authentication methods (pin, fingerprint). Anything I could do to fix this?

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Currently, Authelia does not support platform authenticators, such as Windows Hello, for WebAuthn. This feature is on the roadmap and is expected to be added in a future release, likely in version 4.39. For now, you will not be able to use Windows 11 authentication methods like PIN or fingerprint with Authelia's WebAuthn.

For more details, you can refer to the discussion on the Authelia GitHub page: [WebAuthn with Windows Hello Supported?](https://github.com/authelia/authelia/discussions/5585).

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by james-d-elliott** [comment]
If you're interested there is a build that technically should work with this and it'd be good to get it tested with this specific functionality, it is experimental though and may hose your database so we recommend a backup before delving into it. Let me know.

**Reply by prae82** [reply]
I'm interested! Will this functionality also make it to the stable build soon?

**Reply by james-d-elliott** [reply, marked as answer]
The image is `authelia/authelia:feat-passkeys` and I feel like most of everything I wanted to complete for this release is just in the finalization stage, so probably not too long.

**Comment by prae82** [comment]
I tried the suggested image but when I try to register a device (from my laptop with WIndows 11) I still get this popup only providing a securitykey or mobile. No possibility for Windows PIN or fingerprint.

![image](https://github.com/user-attachments/assets/139588d7-91fc-4614-affe-e63a845998f7)

I tried the same in Proxmox which also has a Webauthn solution. In there it presents a popup with the following information:

![image](https://github.com/user-attachments/assets/9f1957be-84d0-44ac-981a-4349bd633495)

It gives the the option to do Webauthn by Windows PIN or fingerprint.

**Reply by james-d-elliott** [reply]
You may have to set the following new option:

```yaml
webauthn:
  selection_criteria:
    attachment: 'platform'
```

**Reply by prae82** [reply]
Seems like authelia doesn't expect that new key in the config:

```
time="2024-07-17T10:17:14+02:00" level=error msg="Configuration: configuration key not expected: webauthn.selection_criteria.attachment" 

time="2024-07-17T10:17:14+02:00" level=fatal msg="Can't continue due to the errors loading the configuration"
```


I **did** use the new image you mentioned: authelia/authelia:feat-passkeys

**Reply by james-d-elliott** [reply]
Can't replicate that, show the log line indicating the version.

**Reply by prae82** [reply]
I had to uncomment the added 'selection criteria' in the config because only then it shows the version in the log:

```
time="2024-07-17T10:53:54+02:00" level=debug msg="Loaded Configuration Sources" files="[/config/configuration.yml]" filters="[]"
time="2024-07-17T10:53:54+02:00" level=debug msg="Logging Initialized" fields.level=debug file= format= keep_stdout=false
time="2024-07-17T10:53:54+02:00" level=debug msg="Process user information" gid=0 gids="1,2,3,4,6,10,11,20,26,27" name=root uid=0 username=root
time="2024-07-17T10:53:54+02:00" level=info msg="Authelia untagged-v4.38.8 (feat-passkeys, 5ea926f) is starting"
time="2024-07-17T10:53:54+02:00" level=info msg="Log severity set to debug"
time="2024-07-17T10:53:55+02:00" level=info msg="Storage schema is being checked for updates"
time="2024-07-17T10:53:55+02:00" level=info msg="Storage schema is already up to date"
time="2024-07-17T10:53:56+02:00" level=debug msg="Create Server Service (metrics) skipped"
time="2024-07-17T10:53:56+02:00" level=info msg="Startup complete"
time="2024-07-17T10:53:56+02:00" level=info msg="Listening for non-TLS connections on '[::]:9091' path '/'" server=main service=server
```

**Reply by james-d-elliott** [reply]
Ah think maybe the new image never got pushed to the registry you're using, I've re-pushed it.

**Reply by prae82** [reply]
Yes, that did the trick! Thanks!
Apart from the value 'platform', is there a value to support both Windows PIN, fingerprint **and** mobile or Security key?

**Reply by james-d-elliott** [reply]
I don't believe so, the specification only lists two so that's all we've configured, if you see the behavior being different in other providers on the same user agent I can take a closer look though.

**Reply by prae82** [reply]
So you say that it's not possible to let the user decide on what criteria to choose? So it's **or** using the platform (Windows PIN/fingerprint) **or** using the roaming device (mobile)?

This documentation says that there is also a 'no preference' setting?
https://webauthn-doc.spomky-labs.com/pure-php/advanced-behaviours/authenticator-selection-criteria

**Reply by james-d-elliott** [reply]
Looks like if the value is absent that's how it maybe behaves? But it's not entirely clear from the spec. I'll give that a shot, give me a moment.

> If options.[authenticatorSelection](https://www.w3.org/TR/webauthn-2/#dom-publickeycredentialcreationoptions-authenticatorselection) is present:
> 
>     If options.[authenticatorSelection](https://www.w3.org/TR/webauthn-2/#dom-publickeycredentialcreationoptions-authenticatorselection).[authenticatorAttachment](https://www.w3.org/TR/webauthn-2/#dom-authenticatorselectioncriteria-authenticatorattachment) is present and its value is not equal to authenticator’s [authenticator attachment modality](https://www.w3.org/TR/webauthn-2/#authenticator-attachment-modality), [continue](https://infra.spec.whatwg.org/#iteration-continue).

**Reply by james-d-elliott** [reply]
Repull and set it to an empty string.

**Reply by prae82** [reply]
You rock! Works like a charm! Thanks for the support!

![image](https://github.com/user-attachments/assets/707232ed-c7e2-4593-9fb5-19bfab47de8f)

**Comment by prae82** [comment]
Are you using Authelia with Docker? Which version? Because I think the changes are not yet in the main release.

________________________________

**Reply by Smiley83460** [reply]
Hello sorry it works now, the previous replies were hidden so I didn't see the selection_criteria option that need to be added. Thank you

**Comment by prae82** [comment]
Hi @james-d-elliott , I updated the "authelia/authelia:feat-passkeys" image the other day (went from "untagged-v4.38.10" to "untagged-v4.38.16") and found out the Windows Hello wasn't working anymore.

Not sure how it will be merged into the master soon, but it seems the working Windows Hello webauthn was removed somehow in this feat-passkeys branch.

It still works in the v4.38.10 I saved locally.

Btw. Do you know when this Windows Helly webauthn will be in a major release?

**Reply by james-d-elliott** [reply]
Not really any information that's useful here to debugging it. We'd need the same information for a bug report, output of `authelia build-info`, debug logs, and actually providing a description of the issue.

Also did you restore the DB to the pre-beta state before trying?

> Do you know when this Windows Helly webauthn will be in a major release?

I don't give timelines generally unless the release process has started. Also the PR is for passkeys, the windows hello is a side effect but it's not testable.

**Reply by prae82** [reply]
Totally agree and sorry for the brief description! 

But since I'm not that experienced in 'git' I wanted to verify that this "feat-passkeys" is eventually ending up in the main branch. I'm happy to do some more testing if this is the case and if "Windows Hello webauthn" is also in scope.

In the end I would like to use the main branche branch of Authelia ofcourse, IF the "Windows Hello" support is still supposed to be in there.

I was very happy when you did the change and used that image-version until yesterday. But when I updated that image from the hub it stopped working so just wanted to verify this with you.

**Reply by james-d-elliott** [reply]
It theoretically should work, but without any concrete information we're purely guessing.

**Comment by Rick-Phoenix** [comment]
On my windows machine, pin is enabled as a webauthn method, however it does not show up among the choices, both when I leave "attachment" as empty and when I set it to "platform". What am I missing? All other settings are just the defaults from the example configuration in the docs:

```
webauthn:
  disable: false
  enable_passkey_login: true
  display_name: "Authelia"
  attestation_conveyance_preference: "indirect"
  timeout: "60 seconds"
  filtering:
    permitted_aaguids: []
    prohibited_aaguids: []
    prohibit_backup_eligibility: false
  selection_criteria:
    attachment:
    discoverability: "preferred"
    user_verification: "preferred"
  metadata:
    enabled: false
    validate_trust_anchor: true
    validate_entry_permit_zero_aaguid: false
    validate_status: true
    validate_status_permitted: []
    validate_status_prohibited:
      - "REVOKED"
      - "USER_KEY_PHYSICAL_COMPROMISE"
      - "USER_KEY_REMOTE_COMPROMISE"
      - "USER_VERIFICATION_BYPASS"
      - "ATTESTATION_KEY_COMPROMISE"
```

**Reply by prae82** [reply]
Try:

attachment: ""

Works for me.

**Reply by Rick-Phoenix** [reply]
Nope, already tried that too, same result :/
